Modules/Video/Core/include/itkImageToVideoFilter.hxx

```diff
@@ -0,0 +1,265 @@
+{
+  Superclass::PrintSelf(os, indent);
+}
+
+//-PUBLIC METHODS--------------------------------------------------------------
+
+//
+// SetInput( videoStream )
+//
+template <typename TInputImage, typename TOutputVideoStream>
+void
+ImageToVideoFilter<TInputImage, TOutputVideoStream>::SetInput(const TInputImage * image)
+{
+  // We keep this const_cast because in actuality, we do want to be able to
+  // change the requested regions on the input so we need a non-const version
+  this->SetInput(0, const_cast<InputImageType *>(image));
+}
+
+//
+// SetInput( idx, videoStream )
+//
+template <typename TInputImage, typename TOutputVideoStream>
+void
+ImageToVideoFilter<TInputImage, TOutputVideoStream>::SetInput(unsigned int idx, const TInputImage * image)
+{
+  // We keep this const_cast because in actuality, we do want to be able to
+  // change the requested regions on the input so we need a non-const version
+  this->TemporalProcessObject::SetNthInput(idx, const_cast<InputImageType *>(image));
+}
+
+//
+// GetInput()
+//
+template <typename TInputImage, typename TOutputVideoStream>
+const TInputImage *
+ImageToVideoFilter<TInputImage, TOutputVideoStream>::GetInput() const
+{
+  if (this->GetNumberOfInputs() < 1)
+  {
+    return nullptr;
+  }
+  return static_cast<const InputImageType *>(this->ProcessObject::GetInput(0));
+}
+
+//
+// GetInput(idx)
+//
+template <typename TInputImage, typename TOutputVideoStream>
+const TInputImage *
+ImageToVideoFilter<TInputImage, TOutputVideoStream>::GetInput(unsigned int idx) const
+{
+  return static_cast<const InputImageType *>(this->ProcessObject::GetInput(idx));
+}
+
+//-PROTECTED METHODS-----------------------------------------------------------
+
+//
+// GetInput (non-const)
+//
+template <typename TInputImage, typename TOutputVideoStream>
+TInputImage *
+ImageToVideoFilter<TInputImage, TOutputVideoStream>::GetInput()
+{
+  return GetInput(0);
+}
+
+//
+// GetInput (non-const)
+//
+template <typename TInputImage, typename TOutputVideoStream>
+TInputImage *
+ImageToVideoFilter<TInputImage, TOutputVideoStream>::GetInput(unsigned int idx)
+{
+  return static_cast<InputImageType *>(this->ProcessObject::GetInput(idx));
+}
+
+//
+// GenerateOutputInformation
+//
+template <typename TInputImage, typename TOutputVideoStream>
+void
+ImageToVideoFilter<TInputImage, TOutputVideoStream>::GenerateOutputInformation()
+{
+  // Get the input
+  const InputImageType * input = this->GetInput();
+
+  // Get first input frame's largest possible spatial region
+  InputImageRegionType inputRegion = input->GetLargestPossibleRegion();
+
+  // Set temporal frame + duration from user-defined frame axis in input image
+  OutputTemporalRegionType outputTemporalRegion;
+  outputTemporalRegion.SetFrameStart(inputRegion.GetIndex(m_FrameAxis));
+  outputTemporalRegion.SetFrameDuration(inputRegion.GetSize(m_FrameAxis));
+  // TODO set real start and real duration
```

Comment:
Todo?

Reply:
[`itk::TemporalRegion`](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Video/Core/include/itkTemporalRegion.h) can set `RealStart` and `RealDuration` values to represent `RealTimeInterval` timestamps for video frames, but I didn't see a good way to interpret these from a static image input.

Would it make any sense to try to translate these from the "spatial" origin+spacing of the designated temporal axis in the input image, or is it best to leave these uninitialized and request that the user set them manually as available?

Reply:
@thewtex Is there going to be an implicit interpretation in the physical space like LPS is now?  Something like LPST, or LPT?  It looks like by storing the frame index you are half-storing the orientation of the image (i.e., if you change the frame index you change the interpretation of the identity affine matrix).  Is time going to be a first class citizen?

Sorry, I don't have time to review this too much today...  But, is the above example a single frame?  Wouldn't `FrameDuration `be the `GetSpacing(m_FrameAxis)` (i.e. physical spacing) and not the number of pixels in image region?  Similarly, transform the `FrameStart ` index to physical space (don't use `Origin` as the `Index ` can be non-zero)?

Reply:
Good questions! Responding to the second part, `TemporalRegion` represents the region for an entire `TemporalObject`, with `FrameStart` corresponding to the first frame index and `FrameDuration` representing the number of frames in the region.  This may be easiest to see in iteration such as in [`itk::VideoStream`](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Modules/Video/Core/include/itkVideoStream.hxx#L242). 

`RealStart` and `RealDuration` are the corollary to physical origin and spacing, though again I believe `Duration` here corresponds to the time interval across the entire region as opposed to the spacing between frames.

This naming is maybe ambiguous especially when coming from the perspective of how we handle physical spacing in `Image`s. I'll add clarifying comments.

Reply:
If we rename, old names should be available when legacy is on (which is the default). Example use:
```c++
  using ChoicesEnum = ClassNameEnums::Choices;
#if !defined(ITK_LEGACY_REMOVE)
  using Choices = ChoicesEnum;
  static constexpr Choices One = ChoicesEnum::One;
  static constexpr Choices Two = ChoicesEnum::Two;
  static constexpr Choices Three = ChoicesEnum::Three;
#endif
```

Reply:
> @thewtex Is there going to be an implicit interpretation in the physical space like LPS is now? Something like LPST, or LPT?

We are using the order convention here that 3D Slicer is using to be compatible. In any case, we should at least document what is used.

>  Is time going to be a first class citizen?

The `Video` framework makes time a first class citizen :-). The class acts like a bridge into the time is first-class world.

Reply:
@thewtex There isn't currently an order convention baked into `ImageToVideoFilter`. Is it reasonable to expect the user to set an appropriate order with `itk::PermuteAxesImageFilter` prior to generating the `VideoStream`?

Reply:
@dzenanz @brad-t-moore A search returns 161 matching lines making use of `FrameDuration` across the `ITKVideo` module, trying to update that naming would greatly expand the scope of this PR. For now I'm going to add a couple of clarifying comments, we can update naming in a subsequent PR if that would work best.

Reply:
> Is it reasonable to expect the user to set an appropriate order with itk::PermuteAxesImageFilter prior to generating the VideoStream?

Yes, some explanation of the order would help in the documentation, along with a test that demonstrates using `itk::PermuteAxesImagesFilter`.

Reply:
Revisiting above comments to try to clarify where we are here:
> It looks like by storing the frame index you are half-storing the orientation of the image (i.e., if you change the frame index you change the interpretation of the identity affine matrix)

> We are using the order convention here that 3D Slicer is using to be compatible

My understanding is that 3D Slicer conventionally operates on an [RAS coordinate system](https://www.slicer.org/wiki/Coordinate_systems). Do we need to clarify an assumption that image input is in a similar order (most likely RAST)? Or, can we remove this assumption by simply copying over the direction matrix (minus the time axis) to each frame?

To that end, should we allow time to be interpreted with direction such that `TemporalRegion::RealDuration` can be negative? This circles back around to the original `TODO` regarding whether we should attempt to interpret `RealStart` and `RealDuration` from the physical spacing of the temporal axis in the original image.

(Pinging @agirault regarding orientation conventions.)

Reply:
As far as I know, Slicer wants to switch RAS to LPS. But since a long time, when you get an ITK image in Slicer, it is in standard (LPS) orientation. I don't think any conversion is thus needed.

Reply:
Thanks @dzenanz, Brad and I just discussed in person and I believe we agree as well that we should rely on the standard LPS orientation used in ITK.

Some action items on my side to make this happen:
- Document LPS default interpretation for `Image` input and `VideoStream` output frames
- Document and enforce assumption that user-specified temporal axis is axis-aligned (rotating time does not make sense)
- Copy spacing, origin, and direction information from spatial axes in input to output frames
- Interpret spacing, direction and origin information from user-specified temporal axis in input `Image` to inform `RealStart` and `RealDuration` fields in output `VideoStream`

@brad-t-moore Please weigh in if I've left anything off here!

Reply:
Just that @sjh26 dropped by as well and suggested a future `ITKSlicerGlue `or something similar to handle any video conversion weirdness there.

Reply:
Implemented the following assumptions:

> The filter assumes that temporal samples are axis-aligned
> (time cannot be rotated) and the temporal accessor index matches its spatial index (no permutation).

> Other than the specified temporal axis, axis order is preserved and orientation and spacing
  information is copied over to each VideoStream frame from the original Image.
 Image orientation is understood as having LPS (left-posterior-superior) interpretation.

@brad-t-moore Does this cover earlier discussion?

Reply:
@thewtex Per earlier discussion
> some explanation of the order would help in the documentation, along with a test that demonstrates using itk::PermuteAxesImagesFilter.

Currently `ImageToVideoFilter` allows the user to select the input axis to interpret and requires that the physical axis representing time corresponds to the same index as its accessor; i.e., the following is a valid direction matrix for input where time is at column 2:

``` 
[ x x 0
  x x 0
  0 0 1]
```

Are we wanting to handle the case where the accessor index is different from the physical index for the time axis, i.e. where the following is a valid direction with time accessed at index 2?
``` 
[ x x 0
  0 0 1
  x x 0]
```

It looks like `PermuteAxesImageFilter` would not be useful in that case as it also permutes image metadata. If this would be useful we could potentially handle it in `GenerateOutputInformation` with direction matrix column access.

Reply:
> Currently ImageToVideoFilter allows the user to select the input axis to interpret and requires that the physical axis representing time corresponds to the same index as its accessor

:+1: 

> Are we wanting to handle the case where the accessor index is different from the physical index for the time axis, i.e. where the following is a valid direction with time accessed at index 2?

No, I think the current behavior is sufficient. Thanks!